### ISSUE TITLE
Comparing unconstrained generic to null should change state to MaybeDefault

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#47646]
**Version Used**: 3183fa2

```cs
#nullable enable

using System;
public class C {
    static T F1<T>(T t) {
        if (t is null)
        {
            return t; // should warn, but doesn't
        }
        
        t = default;
        return t; // warning
    }
    
    static T F2<T>(T t) where T : class
    {
        if (t is null)
        {
            return t; // warning
        }
        
        t = default;
        return t; // warning
    }
}
```
[SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWAChGABAJgEYWBmCtigYQoBvRhTG92ANgoAVCgDF2AHhkA+ABRyYASmGjxBgJYAzCupgVDUCgWLb9BsSIaPXvAOwUYAbgoB6PwooAAsAeyIUCgB3CARcJApgfAsUULgoXAByGAdHAF9cg0LxCwBeChQ4YwgiH2KxZk8ff0CYuMNcAHNcgpdxXOYpWQVWFQ0tXSjgxEo5EF5WXOc3S1NzS2tbQns+12WV8UavXwDo2NwO7t38+opbsoqqmsI664Mj5tO2i66exjygA=)

**Remarks**: Just because `t` compared equal to `null`, doesn't mean that the domain of `T` allows null.

**Comment by RikkiGibson** [comment]
I think we concluded that this is by-design, because we think that fixing it would introduce unpleasant behavior in other scenarios. It's not possible to "upgrade" a maybe-default value to a maybe-null value via a test, so we're concerned about the effects of letting you "downgrade" a maybe-null value to a maybe-default value via a test.
